{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PROBLEM 1: CONTROLS\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_files(germplasm_file,locusgene_file):\n",
    "    # open the file Germplasm.tsv\n",
    "    germplasm = open(germplasm_file,\"r\")\n",
    "    \n",
    "    # open the file LocusGene.tsv\n",
    "    locusgene = open(locusgene_file,\"r\")\n",
    "    \n",
    "    # These are two empty lists that then will be used\n",
    "    AGI_germplasm = []\n",
    "    AGI_locus = []\n",
    "    \n",
    "    # loop to read the Germplasm file and store every first column entry in a list named AGI_germplasm\n",
    "    for line in germplasm.readlines():\n",
    "        AGI_germplasm.append(line.rstrip().split(\"\\t\")[0])\n",
    "        \n",
    "    # loop to read the LocusGene file and store every first column entry in a list named AGI_locus\n",
    "    for line in locusgene.readlines():\n",
    "        AGI_locus.append(line.rstrip().split(\"\\t\")[0])\n",
    "    \n",
    "    # prints in screen the length of both lists is the same, before I compare them\n",
    "    print(\"The length of the list of AGIs in Germplasm.tsv is: \", str(len(AGI_germplasm)-1), \"and it's the same as LocusGene.tsv: \",str(len(AGI_locus)-1),\"\\n\")\n",
    "    \n",
    "    # loop to prove that both lists are in the same order\n",
    "    for i in range(1,(len(AGI_germplasm))):\n",
    "        \n",
    "        # if the position in AGI_germplasm has the same AGI than that position in AGI_locus\n",
    "        #it prints in the screen the following text with the locus.\n",
    "        if AGI_germplasm[i]==AGI_locus[i]:\n",
    "        \n",
    "            print(i,\"\\t\",\"The AGI in germplasm is: \", AGI_germplasm[i], \"which is the same as the AGI in locus: \", AGI_locus[i])\n",
    "\n",
    "    return\n",
    "\n",
    "read_files(\"Germplasm.tsv\",\"LocusGene.tsv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In order to know if the order followed in both files is the same, I created the function **read_files**. This function recieves both files (Germplasm.tsv and LocusGene.tsv), open them and store all the entries in the first column in two lists, one for each file. These lists stores all the entries with the same order as it appears in the file.\n",
    "\n",
    "Once the lists are completed, I have created a loop that goes from the second element in the list (excluding the first one, because that one is the header of each file) until the total length of the list AGI_germplasm (which length is the same as AGI_locus, as it can be seen printed before the control). Inside the loop, it compares that the AGI in one position of the list germplasm is the same AGI in that position of locusgene's list and if it is true, it is printed in the screen.\n",
    "\n",
    "The results of this script shows that AGI locus are in the same sequence for both files."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PROBLEM 2: Design and create the database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext sql\n",
    "#%config SqlMagic.autocommit=False\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql create database germplasm_locus;\n",
    "%sql use germplasm_locus;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql CREATE TABLE germplasm(locus VARCHAR(9) NOT NULL PRIMARY KEY, germplasm VARCHAR(30) NOT NULL, phenotype VARCHAR(500), pubmed INTEGER NOT NULL)\n",
    "\n",
    "%sql CREATE TABLE locusgene(locus VARCHAR(9) NOT NULL PRIMARY KEY, gene VARCHAR(10) NOT NULL, protein_length INTEGER NOT NULL, FOREIGN KEY (locus) REFERENCES germplasm(locus) ON DELETE CASCADE ON UPDATE CASCADE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To solve this problem I used sqlMagic and to do so, first I used the command to conect to mysql. Then I create the new database \"germplasm_locus\" and select it for its use. \n",
    "\n",
    "After that I create both tables:\n",
    "1. **germplasm**: all of the columns are defined as NOT NULL, so they must have a value when the data is stored.\n",
    "    * locus: it is a combination of 9 numbers and letters, that is why I used a VARCHAR to define it. Also, it is the primary key of the table.\n",
    "    * germplasm: it is also a combination of numbers and letters, so I used another VARCHAR type.\n",
    "    * phenotype: it is a description of the locus, which is formed by sentences. That is why I used the VARCHAR type.\n",
    "    * pubmed: it is a number, so I used the type INTEGER.\n",
    "    \n",
    "    \n",
    "2. **locusgene**: all columns are also defined as NOT NULL, so they must have a value when the data is stored.\n",
    "    * locus: it is the same combination of 9 letters and numbers as in germplasm table, so I used also a VARCHAR. It is also defined as the primary key of this table.\n",
    "    * gene:  it is a combination of numbers and letters with different length, that is why I used a VARCHAR (10), because the length is always less than 10 characters.\n",
    "    * protein_length: it is a number that defines the length of the protein, so it is defined as an INTEGER.\n",
    "    * FOREIGN KEY (locus) REFERENCES germplasm(locus) ON DELETE CASCADE ON UPDATE CASCADE : I used a foreign key that means that locus column in this table is pointing to locus column in the germplasm table, making a 1:1 relationship. The following command means that if an entry is deleted or updated, that entry in germplasm table will also be deleted or updated. I used this foreign key so I could keep the integrity of this database and also if you want to change something, you do not need to do it twice, just doing it once it will be updated to both of the tables."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PROBLEM 3: Fill the database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pymysql.cursors\n",
    "\n",
    "def fill_database(file_to_table):\n",
    "    open_file = open(file_to_table,\"r\")\n",
    "    \n",
    "    #connect to the database\n",
    "    connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='germplasm_locus',\n",
    "                             charset='utf8mb4',  \n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "    try:\n",
    "        with connection.cursor() as cursor:\n",
    "            for line in open_file:\n",
    "                \n",
    "                #with this, I avoid reading the header\n",
    "                # it means that the line must start with an \"A\"\n",
    "                if line.startswith(\"A\"):\n",
    "                    \n",
    "                    #I use the rstrip() to delete \\n and split(\"\\t\") to delete that separation and\n",
    "                    #form a tuple\n",
    "                    line = tuple(line.rstrip().split(\"\\t\"))\n",
    "                    \n",
    "                    #This condition is only true for the Germplasm file\n",
    "                    if (len(line)) == 4 and file_to_table==\"Germplasm.tsv\":\n",
    "                        sql = \"INSERT INTO germplasm (locus,germplasm,phenotype,pubmed) VALUES(%s,%s,%s,%s)\"\n",
    "                        #I use the tuple to insert the information directly because the order the csv file has\n",
    "                        #is the same as the columns formed in mySQL table\n",
    "                        cursor.execute(sql,line)\n",
    "                        connection.commit()\n",
    "                        \n",
    "                    #If the last condition is not true it is because the reading file is Locusgene\n",
    "                    else:\n",
    "                        sql = \"INSERT INTO locusgene(locus,gene,protein_length) VALUES(%s,%s,%s)\"\n",
    "                        #I use the tuple to insert the information directly because the order the csv file has\n",
    "                        #is the same as the columns formed in mySQL tabl\n",
    "                        cursor.execute(sql,line)\n",
    "                        connection.commit()\n",
    "            \n",
    "    finally:\n",
    "        print(\"\")\n",
    "        \n",
    "    return\n",
    "\n",
    "#First I call the function with the Germplasm file\n",
    "fill_database(\"Germplasm.tsv\")\n",
    "#Then I call the function with the Germplasm file\n",
    "fill_database(\"LocusGene.tsv\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To fill the database I created the function **fill_database** that recieves one file at a time. The function opens the file that recieves and then make a connection with the **germplasm_locus** database that I created in problem 2. \n",
    "\n",
    "After that I used a try sentence to fill the tables and I created a loop that reads the file and splits every line, forming a tuple for every line. Then, in order to follow the \"Don't Repeat Yourself\" rule,  I used the next condition: \n",
    "* **if (len(line)) == 4 and file_to_table==\"Germplasm.tsv\"** : this is because I know that the Germplasm.tsv file has 4 columns. But to make sure that is that file, I used the second condition, that means that the file that it's been reading at that moment is \"Germplasm.tsv\". If this is true, the insert is made in the germplasm table.\n",
    "* **else**: if the last condition is not true is because **file_to_table** now is \"LocusGene.tsv\" and I know that its length is 3 columns and not 4 as in germplasm, the locusgene table will be filled.\n",
    "\n",
    "The inserts have something special:\n",
    "\n",
    "* **%s** : this expression include in the values of both inserts, means that the value will be the one that is passed with \"sql\" in cursor.execute. In this case, I include the tuple formed with the line while I was reading the file. I used it because I know that the order of that tuple is the same as the values in the table I created in this database.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PROBLEM 4\n",
    "The new file where the reports are stored is named **reports_exam.tsv** and every report is separated by an empty line.\n",
    "\n",
    "\n",
    "* Create a report that shows the full, joined, content of the two database tables (including a header line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pymysql.cursors\n",
    "\n",
    "# I create the new file in which I will include the results, with the \"write\" mode\n",
    "output=open(\"reports_exam.tsv\",\"w\")\n",
    "\n",
    "# First, the header is printed in the new file\n",
    "output.write(\"locus\\tgermplasm\\tphenotype\\tpubmed\\tlocusgene.locus\\tgene\\tprotein_lenght\\n\")\n",
    "\n",
    "\n",
    "# connect to the database\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='germplasm_locus',\n",
    "                             charset='utf8mb4', \n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        \n",
    "        # this is the selection to obtain the result asked\n",
    "        sql = \"SELECT * FROM germplasm INNER JOIN locusgene ON germplasm.locus = locusgene.locus\"\n",
    "        cursor.execute(sql)\n",
    "        \n",
    "        # I store the results in the variable \"results\"\n",
    "        results = cursor.fetchall()\n",
    "        \n",
    "        #This loops go through the dictionary pymysql shows as the output\n",
    "        for row in results:\n",
    "            #print every value the key dictionary has to the new file\n",
    "            print(row['locus']+\"\\t\"+row['germplasm']+\"\\t\"+row['phenotype']+\"\\t\"+str(row['pubmed'])+\"\\t\"+row[\"locusgene.locus\"]+\"\\t\"+row['gene']+\"\\t\"+str(row['protein_length']), file=output)\n",
    "        \n",
    "        \n",
    "finally:\n",
    "    print(\"\",file=output)\n",
    "    output.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To show both tables joined, I used the SELECT command include in the \"sql\" variable, accesing to the database throw pymysql. Then, in order to store the results, I use the **cursor.fetchall()** with which I could use a loop to read every output as a dictionary. This allows me to include every information to the new file calling the value of every key of the dictionary that shows the result."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Create a joined report that only includes the Genes SKOR and MAA3\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#To include the information in the same file, I open this file I created before with the append mode.\n",
    "output=open(\"reports_exam.tsv\",\"a\")\n",
    "\n",
    "# Connect to the database\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='germplasm_locus',\n",
    "                             charset='utf8mb4', \n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        # this is the selection to obtain the result asked\n",
    "        sql = \"SELECT * FROM germplasm, locusgene WHERE locusgene.gene IN ('SKOR', 'MAA3') AND germplasm.locus = locusgene.locus\"\n",
    "        cursor.execute(sql)\n",
    "        \n",
    "        # I store the results in the variable \"results\"\n",
    "        results = cursor.fetchall()\n",
    "        \n",
    "        #This loops go through the dictionary pymysql shows as the output\n",
    "        for row in results:\n",
    "            \n",
    "            #print every value the key dictionary has to the new file\n",
    "            print(row['locus']+\"\\t\"+row['germplasm']+\"\\t\"+row['phenotype']+\"\\t\"+str(row['pubmed'])+\"\\t\"+row[\"locusgene.locus\"]+\"\\t\"+row['gene']+\"\\t\"+str(row['protein_length']), file=output)\n",
    "        \n",
    "finally:\n",
    "    print(\"\", file=output)\n",
    "    #close the file\n",
    "    output.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I selected the information that was asked and stored it in the same file that I created before and the same way I used before to store it (using cursor.fetchall() and reading the output dictionary).\n",
    "In this report, I did not include a header to the output file because it is the same as the header of the previous report."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Create a report that counts the number of entries for each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pymysql.cursors\n",
    "\n",
    "#To include the information in the same file, I open this file I created before with the append mode.\n",
    "output=open(\"reports_exam.tsv\",\"a\")\n",
    "print(\"Chromosome\\tNumber of entries\", file=output)\n",
    "\n",
    "# Connect to the database\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='germplasm_locus',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "#connection.autocommit = False  # note that it is possible to delay putting changes into the database!\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        # loop that goes from 1 to 5, the chromosomes that we are searching for\n",
    "        for i in range(1,6):\n",
    "            \n",
    "            #search is the variable that contains the pattern I am searching in database\n",
    "            search=\"AT\"+str(i)+\"G%\"\n",
    "            \n",
    "            # this is the selection to obtain the result asked\n",
    "            sql = \"SELECT COUNT(locus) FROM germplasm WHERE locus LIKE %s\"\n",
    "            \n",
    "            # sql is the report to search and the variable \"search\" will fill the \"%s\" in the sql report\n",
    "            cursor.execute(sql,search)\n",
    "            # I store the results in the variable \"results\"\n",
    "            results = cursor.fetchall()\n",
    "            #This loops go through the dictionary pymysql shows as the output\n",
    "            for row in results:\n",
    "                #print every value the key dictionary has to the new file\n",
    "                print(str(i) +\"\\t\"+ str(row['COUNT(locus)']), file= output)\n",
    "        \n",
    "            \n",
    "        \n",
    "finally:\n",
    "    print(\"\",file=output)\n",
    "    #close the file\n",
    "    output.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In order to know how many entries for every chromosome are and follow the \"Don't Repeat Yourself\" rule, I used the following steps:\n",
    "1. **for i in range(1,6)** : I use this loop so it goes from 1 to 5, that are the chromosomes we are searching for.\n",
    "2. **search=\"AT\"+str(i)+\"G%\"** : this variable, search, forms the pattern that is going to be searched in mysql. The patterns starts with \"AT\", then followed by a number that changes from 1 to 5 in the loop and it is forced to be followed by a G and then \"%\", that here means anything any times, so it will find the combination of numbers that follows the G in every AGI Locus.\n",
    "\n",
    "\n",
    "**SELECT COUNT(locus) FROM germplasm WHERE locus LIKE %s**: I used COUNT to count the number of records searched by the query and also I include LIKE in WHERE clause so it will find something that is like the pattern that follows \"LIKE\". I used again \"%s\" in which I include the \"search\" variable (in cursor.execute) that allows to find every chromosome following the loop."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Create a report that shows the average protein length for the genes on each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pymysql.cursors\n",
    "\n",
    "#To include the information in the same file, I open this file I created before with the append mode.\n",
    "output=open(\"reports_exam.tsv\",\"a\")\n",
    "print(\"Chromosome\\tAverage protein length\", file= output)\n",
    "\n",
    "# connect to the database\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='germplasm_locus',\n",
    "                             charset='utf8mb4',  \n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        # this is the selection to obtain the result asked\n",
    "        for i in range(1,6):\n",
    "            \n",
    "            #search is the variable that contains the pattern I am searching in database\n",
    "            search=\"AT\"+str(i)+\"G%\"\n",
    "            # this is the selection to obtain the result asked\n",
    "            sql = \"SELECT AVG(protein_length) FROM locusgene WHERE locus LIKE %s\"\n",
    "            # sql is the report to search and the variable \"search\" will fill the \"%s\" in the sql report\n",
    "            cursor.execute(sql,search)\n",
    "            \n",
    "            # I store the results in the variable \"results\"\n",
    "            results = cursor.fetchall()\n",
    "            \n",
    "            #This loops go through the dictionary pymysql shows as the output\n",
    "            for row in results:\n",
    "                #print every value the key dictionary has to the new file\n",
    "                print( str(i)+\"\\t\"+ str(row['AVG(protein_length)']),file=output)\n",
    "        \n",
    "finally:\n",
    "    print(\"\",file=output)\n",
    "        \n",
    "    output.close()\n",
    "    #close the file\n",
    "    connection.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In order to know how many entries for every chromosome are and follow the \"Don't Repeat Yourself\" rule, I used the following steps:\n",
    "\n",
    "1. **for i in range(1,6)** : I use this loop so it goes from 1 to 5, that are the chromosomes we are searching for.\n",
    "2. **search=\"AT\"+str(i)+\"G%\"** : this variable, search, forms the pattern that is going to be searched in mysql. The pattern starts with \"AT\", then followed by a number that changes from 1 to 5 in the loop and it is forced to be followed by a G and then \"%\", that here means anything any times, so it will find the combination of numbers that follows the G in every AGI Locus.\n",
    "\n",
    "**SELECT AVG(protein_length) FROM locusgene WHERE locus LIKE %s**: I used AVG to know the average protein length of the query searched, using again LIKE in the WHERE clause so it will find the pattern. I also used \"%s\" in which I include the \"search\" variable (in cursor.execute) that allows to find every chromosome following the loop.\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
